go-controller/pkg/networkmanager/nad_controller.go

```diff
@@ -430,3 +528,94 @@ func (nadController *nadController) DoWithLock(f func(network util.NetInfo) erro
 	}
 	return errors.Join(errs...)
 }
+
+// handleNetworkID releases, allocates and annotates network IDs. If both old
+// and new networks are provided, it is assumed they are a different network
+// with the same name and will be allocated the same ID.
+func (c *nadController) handleNetworkID(old util.NetInfo, new util.MutableNetInfo, nad *nettypes.NetworkAttachmentDefinition) error {
+	if new != nil && new.GetNetworkName() == types.DefaultNetworkName {
+		return nil
+	}
+
+	var err error
+	var id int
+
+	// check what ID is currently annotated
+	if nad != nil && nad.Annotations[types.OvnNetworkIDAnnotation] != "" {
+		annotated := nad.Annotations[types.OvnNetworkIDAnnotation]
+		id, err = strconv.Atoi(annotated)
+		if err != nil {
+			return fmt.Errorf("failed to parse annotated network ID: %w", err)
+		}
+	}
+
+	// this is not the cluster manager nad controller and we are not allocating
+	// so just return what we got from the annotation
+	if c.networkIDAllocator == nil {
+		if new != nil {
+			new.SetNetworkID(id)
+		}
+		return nil
+	}
+
+	// release old ID
+	if old != nil && new == nil && old.GetNetworkName() != types.DefaultNetworkName {
+		c.networkIDAllocator.ReleaseID(old.GetNetworkName())
+		return nil
+	}
+
+	// nothing to allocate
+	if new == nil {
+		return nil
+	}
+	name := new.GetNetworkName()
+
+	// an ID was annotated, check if it is free to use or stale
+	if id != 0 {
+		err = c.networkIDAllocator.ReserveID(name, id)
+		if err != nil {
+			// already reserved for a different network, allocate a new id
+			id = 0
```

Comment:
That's used when we can't parse from annotation, but not used by the allocator itself.

What do you mean with `doing negative checks with the default network ID`?

Reply:
i mean you are setting id here to 0 to indicate that you  dont have an id, so in the if statement below you will use that to know you need to allocate a new one. This is not good cause 0 is actually a valid and reserved ID. It would be better to use a constant. Since InvalidID is -1, but that one is not as good since it means something else, maybe you could use -2 as NoID. Alternatively, could use a pointer and nil means no value.

Reply:
I don't see the actual problem since this function is not doing anything for the default network and locally it is well known that 0 is an invalid ID for anything that is not the default network, since that is also being allocated here. 
But I see this is confusing so I will use that invalid ID constant.